crates/ra_assists/src/fill_struct_fields.rs

```diff
@@ -0,0 +1,120 @@
+use std::fmt::Write;
+
+use hir::{AdtDef, Ty, source_binder};
+use hir::db::HirDatabase;
+
+use ra_syntax::ast::{self, AstNode, Expr};
+
+use crate::{AssistCtx, Assist, AssistId};
+
+pub(crate) fn fill_struct_fields(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let struct_lit = ctx.node_at_offset::<ast::StructLit>()?;
+
+    // If we already have existing struct fields, don't provide the assist.
+    // TODO: provide assist for tuple structs
+    match struct_lit.named_field_list() {
+        Some(named_field_list) if named_field_list.fields().count() > 0 => {
+            return None;
+        }
+        _ => {}
+    }
+
+    let expr: &Expr = struct_lit.into();
+    let function =
+        source_binder::function_from_child_node(ctx.db, ctx.frange.file_id, struct_lit.syntax())?;
+
+    let infer_result = function.infer(ctx.db);
+    let source_map = function.body_source_map(ctx.db);
+    let node_expr = source_map.node_expr(expr)?;
+    let struct_lit_ty = infer_result[node_expr].clone();
+    let struct_def = match struct_lit_ty {
+        Ty::Adt { def_id: AdtDef::Struct(s), .. } => s,
+        _ => return None,
+    };
+
+    let struct_name = struct_def.name(ctx.db)?;
+    let db = ctx.db;
+
+    ctx.add_action(AssistId("fill_struct_fields"), "fill struct fields", |edit| {
+        let mut buf = format!("{} {{\n", struct_name);
```

Comment:
I don't think we should replace struct literal as a whole. For example, it could use `Self` instead of the literal name of the struct, and we shouldn't change that. We should replace `NamedFieldList`.